### PR TITLE
Allow Start Last Run in the Overview to work again

### PR DESCRIPTION
**Pull request by andrewballantyne**
https://jira.coreos.com/browse/ODC-2177

Fixes the Start Last Run button on the Overview pane in Topology. It ran into a conflict with another work that reworked the API.

Issue:
StartLastRun

Fixed:
StartLastRunWorking

The shaking appears related to https://jira.coreos.com/browse/ODC-2122 if you look at the dev tools you can see the vibration happens at the time it remounts.

### CODE REVIEW

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineRunItem.tsx [hunk 1]
```diff
@@ -30,7 +30,7 @@ const PipelineRunItem: React.FC<PipelineRunItemProps> = ({ pipelineRun }) => {
           {lastUpdated && <span className="text-muted">&nbsp;({fromNow(lastUpdated)})</span>}
         </GridItem>
         <GridItem span={3}>
-          <Status status={pipelineRunFilterReducer(pipelineRun)} />
+          <Status status={pipelineRunStatus(pipelineRun) || 'Pending'} />
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineRunItem.tsx]: Rare instances it wouldn't start immediately and showed `-` which I felt could be better shown as `Pending`.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx [hunk 2]
```diff
@@ -1,30 +1,29 @@
...
   impersonate,
-  disabled,
 }) => {
-  const latestRun = _.get(pipeline, ['latestRun'], null);
-  const { label, callback, accessReview } = rerunPipeline(PipelineModel, pipeline, latestRun);
+  const latestRun = getLatestRun({ data: pipelineRuns }, 'startTimestamp');
+  const { label, callback, accessReview } = rerunPipeline(PipelineModel, latestRun, null, {
+    pipelineRunRedirect: false,
+  });
   const isAllowed = useAccessReview(accessReview, impersonate);
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: This probably shouldn't be a hook. `accessReview` is optional but it *must* call this function every time due to the way hooks work.

**Comment by andrewballantyne** [comment]
/assign @sahil143 

This is fixing of your feature work during the merge conflicts that make this call no longer work. Please review and let me know if you think the heart of what you had is still here.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 3]
```diff
@@ -131,13 +131,36 @@ export const startPipeline: KebabAction = (
   },
 });
 
-export const rerunPipeline: KebabAction = (kind: K8sKind, pipelineRun: PipelineRun) => {
-  // The returned function will be called using the 'kind' and 'obj' in Kebab Actions
+type RerunPipelineData = {
+  pipelineRunRedirect?: boolean;
+};
+export const rerunPipeline: KebabAction = (
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: I don't like this change, so I am simplifying it.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 4]
```diff
@@ -131,13 +131,36 @@ export const startPipeline: KebabAction = (
...
+) => {
+  const { pipelineRunRedirect = true } = customData;
+
+  const sharedProps = { label: 'Start Last Run', accessReview: {} };
+
+  if (
+    !pipelineRun ||
+    !_.has(pipelineRun, 'metadata.name') ||
+    !_.has(pipelineRun, 'metadata.namespace')
+  ) {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: Why are you checking for metadata.name and namespace here? If `pipelineRun` is there then metadata should be there.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 5]
```diff
@@ -131,13 +131,36 @@ export const startPipeline: KebabAction = (
...
 
-export const rerunPipeline: KebabAction = (kind: K8sKind, pipelineRun: PipelineRun) => {
-  // The returned function will be called using the 'kind' and 'obj' in Kebab Actions
+type RerunPipelineData = {
+  pipelineRunRedirect?: boolean;
+};
+export const rerunPipeline: KebabAction = (
+  kind: K8sKind,
+  pipelineRun: PipelineRun,
+  resources: any,
```
Comment (sahil143) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: resources is not being used in the method

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 6]
```diff
@@ -131,13 +131,36 @@ export const startPipeline: KebabAction = (
...
+  customData: RerunPipelineData = {},
```
Comment (sahil143) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: We can define {pipelineRunRedirect: true} here and make it optional.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 7]
```diff
@@ -131,13 +131,32 @@ export const startPipeline: KebabAction = (
   },
 });
 
-export const rerunPipeline: KebabAction = (kind: K8sKind, pipelineRun: PipelineRun) => {
-  // The returned function will be called using the 'kind' and 'obj' in Kebab Actions
+type RerunPipelineData = {
+  onComplete?: (pipelineRun: PipelineRun) => void;
+};
+const rerunPipeline: KebabAction = (
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: I logged https://jira.coreos.com/browse/ODC-2184 to address the fact that this function was converted to being a rereun pipeline runs.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx [hunk 8]
```diff
@@ -150,6 +169,14 @@ export const rerunPipeline: KebabAction = (kind: K8sKind, pipelineRun: PipelineR
...
 };
 
+export const rerunPipelineAndStay: KebabAction = (kind: K8sKind, pipelineRun: PipelineRun) => {
+  return rerunPipeline(kind, pipelineRun);
+};
+
+export const rerunPipelineAndRedirect: KebabAction = (kind: K8sKind, pipelineRun: PipelineRun) => {
+  return rerunPipeline(kind, pipelineRun, null, { onComplete: handlePipelineRunSubmit });
+};
+
```
Comment (sahil143) [inline frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: Why this change? I think flag thing (pipelineRunRedirect) worked fine.
You are still adding a check
`typeof onComplete === 'function' ? onComplete : () => {}`

Comment (sahil143) [reply frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: By default we can keep the flag true or false depending on majority usage of flag and make `customData` optional. WDYT?

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx (hunk 4 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: Not according to the type. `PipelineRun` doesn't define it, but extends `K8sResourceKind` and it specifies it as optional. Of which inside `metadata`, `name` and `namespace` are also optional.

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx (hunk 5 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: Nope, but it's the signature of the `KebabAction`, and I need the `customData` for my callbacks:
```
export type KebabAction = (
  kind: K8sKind,
  obj: K8sResourceKind,
  resources?: any,
  customData?: any,
) => KebabOption;
```

File: frontend/packages/dev-console/src/utils/pipeline-actions.tsx (hunk 4 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/pipeline-actions.tsx]: @sahil143 unless the type we are currently using says this, I won't write code that "assumes" optional params are there. Types should not be just ignored when it's inconvenient. This code is currently only executed on click so it's not a render-cycle slowdown.

**Comment by sahil143** [comment]
/lgtm

verified changes locally. It works!!
@andrewballantyne just wanted to make a note regarding the "mount and unmount issue with pipeline component" and we already have a bug for this (https://jira.coreos.com/browse/ODC-2122) but it has become noticeable on the UI.
mu

**Comment by andrewballantyne** [comment]
> @andrewballantyne just wanted to make a note regarding the "mount and unmount issue with pipeline component" and we already have a bug for this (https://jira.coreos.com/browse/ODC-2122) but it has become noticeable on the UI.

Yup :) I posted that in my description of the PR.
